In [10]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [13]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(5)

price_change_percentage_24h  price_change_percentage_7d   
coin_id                                                                 
bitcoin                           1.08388                     7.60278  \
ethereum                          0.22392                    10.38134   
tether                           -0.21173                     0.04935   
ripple                           -0.37819                    -0.60926   
bitcoin-cash                      2.90585                    17.09717   

              price_change_percentage_14d  price_change_percentage_30d   
coin_id                                                                  
bitcoin                           6.57509                      7.67258  \
ethereum                          4.80849                      0.13169   
tether                            0.00640                     -0.04237   
ripple                            2.24984                      0.23455   
bitcoin-cash                     14.75334                     15.74903   

              price_change_percentage_60d  price_change_percentage_200d   
coin_id                                                                   
bitcoin                          -3.25185                      83.51840  \
ethereum                        -12.88890                     186.77418   
tether                            0.28037                      -0.00542   
ripple                          -17.55245                      39.53888   
bitcoin-cash                    -13.71793                      21.66042   

              price_change_percentage_1y  
coin_id                                   
bitcoin                         37.51761  
ethereum                       101.96023  
tether                           0.01954  
ripple                         -16.60193  
bitcoin-cash                    14.49384

In [14]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d   
count                    41.000000                   41.000000  \
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d   
count                    41.000000                    41.000000  \
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d   
count                    41.000000                     41.000000  \
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [15]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [16]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
data = StandardScaler().fit_transform(df_market_data.values)

In [17]:
# Create a DataFrame with the scaled data
data_df = pd.DataFrame(data, index=df_market_data.index, columns=df_market_data.columns)

# Copy the crypto names from the original data

# Set the coinid column as index


# Display sample data
data_df.head(5)

price_change_percentage_24h  price_change_percentage_7d   
coin_id                                                                 
bitcoin                          0.508529                    0.493193  \
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d   
coin_id                                                                  
bitcoin                          0.772200                     0.235460  \
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d   
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953  \
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [18]:
# Create a list with the number of k-values from 1 to 11
k = [x for x in range(1, 11)]

In [19]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
inertia = [KMeans(n_clusters=i, random_state=1).fit(data_df).inertia_ for i in k]


c:\Users\Daniel Yoon\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\Daniel Yoon\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\Daniel Yoon\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\Daniel Yoon\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto

In [21]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    'k': k,
    'inertia': inertia
}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df = pd.DataFrame(elbow_data)

elbow_df.head(5)

k     inertia
0  1  287.000000
1  2  195.820218
2  3  123.190482
3  4   79.022435
4  5   65.405923

In [22]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_df.hvplot.line(
    x="k", 
    y="inertia", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [23]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [25]:
# Fit the K-Means model using the scaled data
model.fit(data_df)

c:\Users\Daniel Yoon\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [26]:
# Predict the clusters to group the cryptocurrencies using the scaled data
prediction = model.predict(data_df)

# Print the resulting array of cluster values.
prediction

array([3, 3, 1, 1, 3, 3, 3, 3, 3, 1, 1, 1, 1, 3, 1, 3, 1, 1, 3, 1, 1, 3,
       1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 0, 3, 1, 1, 2, 1, 1, 1, 1])

In [28]:
# Create a copy of the DataFrame
prediction_df = data_df.copy()

prediction_df.head(5)

price_change_percentage_24h  price_change_percentage_7d   
coin_id                                                                 
bitcoin                          0.508529                    0.493193  \
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d   
coin_id                                                                  
bitcoin                          0.772200                     0.235460  \
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d   
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953  \
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

In [29]:
# Add a new column to the DataFrame with the predicted clusters
prediction_df['clusters'] = prediction

# Display sample data
prediction_df.head(5)

price_change_percentage_24h  price_change_percentage_7d   
coin_id                                                                 
bitcoin                          0.508529                    0.493193  \
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d   
coin_id                                                                  
bitcoin                          0.772200                     0.235460  \
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d   
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953  \
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  clusters  
coin_id                                             
bitcoin                        -0.251637         3  
ethereum                       -0.199352         3  
tether                         -0.282061         1  
ripple                         -0.295546         1  
bitcoin-cash                   -0.270317         3

In [30]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
prediction_df.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="clusters",
    hover_cols="coin_id"
)

:NdOverlay   [clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [31]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [32]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
pca_data = pca.fit_transform(data_df) 

# View the first five rows of the DataFrame. 
pca_data_df = pd.DataFrame(pca_data)
pca_data_df.head(5)

0         1         2
0 -0.600667  0.842760  0.461595
1 -0.458261  0.458466  0.952877
2 -0.433070 -0.168126 -0.641752
3 -0.471835 -0.222660 -0.479053
4 -1.157800  2.041209  1.859715

In [33]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_, pca.explained_variance_ratio_

(array([2.6689967 , 2.48978331, 1.26307212]),
 array([0.3719856 , 0.34700813, 0.17603793]))

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The first component accounts for 2.67 units, which represents 37% of the total variance. Similarly, the second component explains 2.49 units, corresponding to 35% of the variance. The third component contributes 1.26 units, explaining 18% of the variance. In total, these three components together explain 6.42 units, which accounts for 89.5% of the total variance.

In [35]:
# Create a new DataFrame with the PCA data.

# Creating a DataFrame with the PCA data
# Copy the crypto names from the original data
# Set the coinid column as index
pca_df = pd.DataFrame(
    pca_data, 
    index=df_market_data.index, 
    columns=["PCA1", "PCA2", "PCA3"])

# Display sample data
pca_df.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [36]:
# Create a list with the number of k-values from 1 to 11
pca_k = [x for x in range(1, 11)]

In [38]:
# Create an empy list to store the inertia values
pca_inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in pca_k:
    model = KMeans(n_clusters=i, random_state=1)
    model.fit(pca_df)
    pca_inertia.append(model.inertia_)

c:\Users\Daniel Yoon\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\Daniel Yoon\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\Daniel Yoon\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\Daniel Yoon\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto

In [40]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {
    'k': pca_k,
    'inertia': pca_inertia
}

# Create a DataFrame with the data to plot the Elbow curve
elbow_pca_df = pd.DataFrame(elbow_data_pca)

elbow_pca_df.head(5)

k     inertia
0  1  256.874086
1  2  165.901994
2  3   93.774626
3  4   49.665497
4  5   38.352251

In [41]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_pca_df.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=pca_k
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, it doesn't

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [42]:
# Initialize the K-Means model using the best value for k
model_pca = KMeans(n_clusters=4, random_state=1)

In [43]:
# Fit the K-Means model using the PCA data
model_pca.fit(pca_df)

c:\Users\Daniel Yoon\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [44]:
# Predict the clusters to group the cryptocurrencies using the PCA data
pca_prediction = model_pca.predict(pca_df)

# Print the resulting array of cluster values.
pca_prediction

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 0, 3, 0, 0, 0, 0])

In [45]:
# Create a copy of the DataFrame with the PCA data
pca_prediction_df = pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
pca_prediction_df["clusters_pca"] = pca_prediction

# Display sample data
pca_prediction_df.head(5)

PCA1      PCA2      PCA3  clusters_pca
coin_id                                                 
bitcoin      -0.600667  0.842760  0.461595             1
ethereum     -0.458261  0.458466  0.952877             1
tether       -0.433070 -0.168126 -0.641752             0
ripple       -0.471835 -0.222660 -0.479053             0
bitcoin-cash -1.157800  2.041209  1.859715             1

In [46]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
pca_prediction_df.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by="clusters_pca",
    hover_cols="coin_id"
)

:NdOverlay   [clusters_pca]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [48]:
# Composite plot to contrast the Elbow curves

elbow_df.hvplot.line(
    x="k", 
    y="inertia", 
    xticks=k,
    label="Original data"
) * elbow_pca_df.hvplot.line(
    x="k", 
    y="inertia", 
    xticks=pca_k,
    label="PCA data"
)

:Overlay
   .Curve.Original_data :Curve   [k]   (inertia)
   .Curve.PCA_data      :Curve   [k]   (inertia)

In [49]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!
prediction_df.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="clusters",
    hover_cols="coin_id",
    label="k-means Original data"
) + pca_prediction_df.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by="clusters_pca",
    hover_cols="coin_id",
    label="k-means PCA data"
)

:Layout
   .NdOverlay.K_hyphen_minus_means_Original_data :NdOverlay   [clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.K_hyphen_minus_means_PCA_data      :NdOverlay   [clusters_pca]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:**  Applying PCA analysis reduces the dataset's inertia and enhances the visual clarity of distinct clusters.